File: src/class071/C5_ReverseArraySubarrayMaxSum.py
from typing import List
import random

def reverse(nums: List[int], l: int, r: int) -> None:
    # nums[l...r]范围上的数字进行逆序调整
    while l < r:
        nums[l], nums[r] = nums[r], nums[l]
        l += 1
        r -= 1

def max_sum(nums: List[int]) -> int:
    # 返回子数组最大累加和
    n = len(nums)
    ans = pre = nums[0]
    for i in range(1, n):
        pre = max(nums[i], pre + nums[i])
        ans = max(ans, pre)
    return ans

def max_sum_reverse1(nums: List[int]) -> int:
    # 暴力方法，为了验证
    ans = float('-inf')
    for l in range(len(nums)):
        for r in range(l, len(nums)):
            nums_copy = nums[:]
            reverse(nums_copy, l, r)
            ans = max(ans, max_sum(nums_copy))
    return ans

def max_sum_reverse2(nums: List[int]) -> int:
    # 正式方法，时间复杂度O(n)
    n = len(nums)
    start = [0] * n
    start[-1] = nums[-1]
    for i in range(n-2, -1, -1):
        start[i] = max(nums[i], nums[i] + start[i + 1])
    ans = start[0]
    end = max_end = nums[0]
    for i in range(1, n):
        ans = max(ans, max_end + start[i])
        end = max(nums[i], end + nums[i])
        max_end = max(max_end, end)
    return max(ans, max_end)

def random_array(n: int, v: int) -> List[int]:
    # 生成随机数组
    return [random.randint(-v, v) for _ in range(n)]

# 为了测试
if __name__ == "__main__":
    n = 50
    v = 200
    test_time = 20000
    print("测试开始")
    for _ in range(test_time):
        len_ = random.randint(1, n)
        arr = random_array(len_, v)
        ans1 = max_sum_reverse1(arr)
        ans2 = max_sum_reverse2(arr)
        if ans1 != ans2:
            print("出错了!")
    print("测试结束")

File: src/class071/C5_ReverseArraySubarrayMaxSum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <cstdlib>  // for rand()
#include <ctime>    // for time()

using namespace std;

class Solution {
public:
    // Function to reverse a part of the array
    void reverse(vector<int>& nums, int l, int r) {
        while (l < r) {
            swap(nums[l], nums[r]);
            l++;
            r--;
        }
    }

    // Function to find the maximum sum of a subarray
    int maxSum(vector<int>& nums) {
        int n = nums.size();
        int ans = nums[0], pre = nums[0];
        for (int i = 1; i < n; i++) {
            pre = max(nums[i], pre + nums[i]);
            ans = max(ans, pre);
        }
        return ans;
    }

    // Brute-force method to verify
    int maxSumReverse1(vector<int>& nums) {
        int ans = INT_MIN;
        for (int l = 0; l < nums.size(); l++) {
            for (int r = l; r < nums.size(); r++) {
                reverse(nums, l, r);
                ans = max(ans, maxSum(nums));
                reverse(nums, l, r);  // Reverse back to original state
            }
        }
        return ans;
    }

    // Optimized method with O(n) time complexity
    int maxSumReverse2(vector<int>& nums) {
        int n = nums.size();
        vector<int> start(n);
        start[n - 1] = nums[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            start[i] = max(nums[i], nums[i] + start[i + 1]);
        }
        int ans = start[0];
        int end = nums[0], maxEnd = nums[0];
        for (int i = 1; i < n; i++) {
            ans = max(ans, maxEnd + start[i]);
            end = max(nums[i], end + nums[i]);
            maxEnd = max(maxEnd, end);
        }
        ans = max(ans, maxEnd);
        return ans;
    }
};

// Function to generate a random array
vector<int> randomArray(int n, int v) {
    vector<int> ans(n);
    for (int i = 0; i < n; i++) {
        ans[i] = rand() % (v * 2 + 1) - v;
    }
    return ans;
}

// Main function to test the methods
int main() {
    srand(time(nullptr));  // Initialize random number generator
    int n = 50;
    int v = 200;
    int testTime = 20000;
    cout << "测试开始" << endl;
    Solution sol;
    for (int i = 0; i < testTime; i++) {
        int len = rand() % n + 1;
        vector<int> arr = randomArray(len, v);
        int ans1 = sol.maxSumReverse1(arr);
        int ans2 = sol.maxSumReverse2(arr);
        if (ans1 != ans2) {
            cout << "出错了!" << endl;
        }
    }
    cout << "测试结束" << endl;
    return 0;
}
